### ISSUE TITLE
Installation issue: lbann@0.101

### ISSUE THREAD

**Issue by tldahlgren** [LLNL/spack#23213]
I am currently building an environment -- concretize separately -- on LC's quartz system.

The specification provided below is the one produced by the clingo concretizer plus an additional constraints mpi to reproduce the output outside of my environment.

@bvanessen 

```console
$ spack install lbann@0.101%gcc@8.3.1+al~boost+conduit~cuda~deterministic+dihydrogen~distconv~docs~extras~fft~half+hwloc~ipo~numpy~nvprof~nvshmem~onednn+pfe+python~rocm~vision~vtune amdgpu_target=none build_type=Release cuda_arch=none dtype=float arch=linux-rhel7-broadwell ^mvapich2@2.3.5
...
==> lbann: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Ninja' '-DCMAKE_INSTALL_PREFIX:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/lbann-0.101-46apqavf4mf6cdymy4nf25o6lxktqmey' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/lbann-0.101-46apqavf4mf6cdymy4nf25o6lxktqmey/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/lbann-0.101-46apqavf4mf6cdymy4nf25o6lxktqmey/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hwloc-1.11.13-fcnglt7stspn55w3crqp7zhd4kyalext/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/libpciaccess-0.16-qdqbgcm5wqaqco3omfiagc7vubztwr2d/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/libxml2-2.9.10-ekqeeh6cngvdhuccnqlllmtlgupnzetm/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/libiconv-1.16-ntrmibdp7r4xx7iwx7pyjmriu242i7wq/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/zlib-1.2.11-c7zv5ftnxfe4h3237t42mshbrwd4ecv7/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/mvapich2-2.3.5-tykec5h34psoihqxacocdtxfjnd7ozxi/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/rdma-core-34.0-llmzjnskah47tnvcymrknigcamrwbimx/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/libnl-3.3.0-cbtknyw7adogpeaajchy2wli2h3ygqbt/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/conduit-develop-w6borpjmepadojnfwn7f4ptj2j3iiutf/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hdf5-1.10.7-n232bnukq57xus5hqz744lozpqboqays/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.14-o4l6rdih7caiywq4f3phejwkvvhmkugl/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hydrogen-1.4.0-33sjeb3aymz6a77vn4wn5pf42yyfvggt/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/aluminum-0.4.0-al3ydedigl23ihcntz2e7buhbsccbviz/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/rdma-core-34.0-llmzjnskah47tnvcymrknigcamrwbimx/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dihydrogen-0.1-qnmh7rjrlbvfmrvk6lzo6zchs4ak34fc/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/protobuf-3.10.0-r5qzsptjduhg327pnbjlarpsq2j6caxj/lib64' '-DCMAKE_PREFIX_PATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/aluminum-0.4.0-al3ydedigl23ihcntz2e7buhbsccbviz;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/catch2-2.13.4-z5hgakm6om24sq6wz3i7gffvxmoim4k6;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cereal-1.3.0-tjda6ery4zxycccjmf4wturdw3fgttnz;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/clara-1.1.5-va3d4rnv74evsz2jus7qtgtmhcfr2l2q;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cmake-3.20.1-t5tqx56iqdkurjwg7sbl2kuw2w5j7wuo;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/conduit-develop-w6borpjmepadojnfwn7f4ptj2j3iiutf;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dihydrogen-0.1-qnmh7rjrlbvfmrvk6lzo6zchs4ak34fc;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hwloc-1.11.13-fcnglt7stspn55w3crqp7zhd4kyalext;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hydrogen-1.4.0-33sjeb3aymz6a77vn4wn5pf42yyfvggt;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/mvapich2-2.3.5-tykec5h34psoihqxacocdtxfjnd7ozxi;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/ninja-1.10.2-3oesdmmu5slmav2wmvchtcspug4nsigr;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/protobuf-3.10.0-r5qzsptjduhg327pnbjlarpsq2j6caxj;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/py-protobuf-3.10.0-jpkvyoqe44ixpxjp5seegphtajwwagrp;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/py-setuptools-50.3.2-zupmheywzqvb3stvcanbo7lo3m3tqv7w;/usr/tce;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/zstr-1.0.4-yu6cnbs7o5kzprcn27rpwzb2fn5sn652' '-DCMAKE_CXX_FLAGS=-DLBANN_SET_EL_RNG' '-DLBANN_VERSION=spack' '-DCMAKE_CXX_STANDARD=17' '-DLBANN_WITH_CNPY=False' '-DLBANN_DETERMINISTIC:BOOL=False' '-DLBANN_WITH_HWLOC=True' '-DLBANN_WITH_ALUMINUM:BOOL=True' '-DLBANN_WITH_BOOST:BOOL=False' '-DLBANN_WITH_CONDUIT:BOOL=True' '-DLBANN_WITH_NVSHMEM:BOOL=False' '-DLBANN_WITH_FFT:BOOL=False' '-DLBANN_WITH_ONEDNN:BOOL=False' '-DLBANN_WITH_EMBEDDED_PYTHON:BOOL=True' '-DLBANN_WITH_PYTHON_FRONTEND:BOOL=True' '-DLBANN_WITH_TBINF=OFF' '-DLBANN_WITH_UNIT_TESTING:BOOL=False' '-DLBANN_WITH_VISION:BOOL=False' '-DLBANN_WITH_VTUNE:BOOL=False' '-DLBANN_DATATYPE=float' '-DCEREAL_DIR=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cereal-1.3.0-tjda6ery4zxycccjmf4wturdw3fgttnz' '-DProtobuf_DIR=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/protobuf-3.10.0-r5qzsptjduhg327pnbjlarpsq2j6caxj' '-Dprotobuf_MODULE_COMPATIBLE=ON' '-DHydrogen_DIR=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/hydrogen-1.4.0-33sjeb3aymz6a77vn4wn5pf42yyfvggt/CMake/hydrogen' '-DAluminum_DIR=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/aluminum-0.4.0-al3ydedigl23ihcntz2e7buhbsccbviz' '-DConduit_DIR=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/conduit-develop-w6borpjmepadojnfwn7f4ptj2j3iiutf' '-DLBANN_WITH_DIHYDROGEN:BOOL=True' '-DLBANN_WITH_DISTCONV:BOOL=False' '/var/tmp/dahlgren/spack-stage/spack-stage-lbann-0.101-46apqavf4mf6cdymy4nf25o6lxktqmey/spack-src'

1 error found in build log:
     106    -- Looking for openblas_get_num_procs - found
     107    -- Found BLASImpl: /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.14-o4l6rdih7caiywq4f3phejwkvvhmkugl/lib/libopenblas.so
     108    -- Found DiHydrogen components: Core;Meta;Patterns
     109    -- Aluminum found in Hydrogen. Using Aluminum.
     110    -- Found ZLIB: /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/zlib-1.2.11-c7zv5ftnxfe4h3237t42mshbrwd4ecv7/lib/libz.so (found version
             "1.2.11")
     111    -- Found Protobuf: /usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/protobuf-3.10.0-r5qzsptjduhg327pnbjlarpsq2j6caxj/lib64/cmake/protob
            uf
  >> 112    CMake Error at CMakeLists.txt:249 (find_package):
     113      Could not find a package configuration file provided by "OpenCV" with any
     114      of the following names:
     115    
     116        OpenCVConfig.cmake
     117        opencv-config.cmake
```

### Information on your system

* **Spack:** 0.16.1-2245-4cca69f4ed
* **Python:** 3.7.2
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo

### Additional information

* spack-build-out.txt does not add value over the output above.
* spack-build-env.txt is empty
* [CMakeOutput.log](https://github.com/spack/spack/files/6361811/lbann-CMakeOutput.log)
* [CMakeError.log](https://github.com/spack/spack/files/6361812/lbann-CMakeError.log)

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files or more relevant CMake log files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by bvanessen** [comment]
@tldahlgren How recently have you updated your spack repo?  On Tuesday we put together a series of PRs that fixed some of these bugs, specifically, that the `+vision` variant has to be turned on for the older version of lbann.

**Comment by tldahlgren** [comment]
> @tldahlgren How recently have you updated your spack repo? On Tuesday we put together a series of PRs that fixed some of these bugs, specifically, that the `+vision` variant has to be turned on for the older version of lbann.

Looks like it was last Friday (April 16, 2021).

**Comment by bvanessen** [comment]
@tldahlgren Can you please update and check again.

**Comment by tldahlgren** [comment]
> @tldahlgren Can you please update and check again.

Sure.  Will try again once my latest environment-based build completes (IF the package doesn't build after the update). :grin:

**Comment by alalazo** [comment]
@adamjstewart FYI, this _may_ be related to #22340

**Comment by adamjstewart** [comment]
I think this is a bug in the `lbann` package. The spec above explicitly specifies `~vision`, so `lbann` does not depend on `opencv` at all.

**Comment by bvanessen** [comment]
@adamjstewart @tldahlgren Version 0.101.0 of LBANN did not support disabling the dependency on OpenCV.  That was only added in the develop branch.

**Comment by adamjstewart** [comment]
In that case we can just add a conflict for `@:0.101 ~vision`

**Comment by bvanessen** [comment]
That line is already in there `conflicts('~vision', when='@0.91:0.101')`

**Comment by adamjstewart** [comment]
Ah, so it sounds like this issue may have been fixed by #23159. @tldahlgren can you try updating to develop again?

**Comment by tldahlgren** [comment]
@bvanessen @adamjstewart @alalazo Thank you all for the exchange and information.

I'm closing this issue since I am now able to build `lbann` (as part of an environment) after updating to develop.
